ci/build-test-matrix.js

```diff
@@ -5,154 +5,241 @@
-    "gcc_package": "gcc-riscv64-linux-gnu",
-    "gcc": "riscv64-linux-gnu-gcc",
-    "qemu": "qemu-riscv64 -cpu rv64,v=true,vlen=256,vext_spec=v1.0,Zfa=true,zba=true,zbb=true,zbc=true,zbs=true,zbkb=true,zcb=true -L /usr/riscv64-linux-gnu",
-    "qemu_target": "riscv64-linux-user",
-    "name": "Test Linux riscv64",
-    "filter": "linux-riscv64",
-    "isa": "riscv64",
-  }
-];
-
-for (let config of array) {
-  if (config.rust === undefined) {
-    config.rust = 'default';
-  }
+async function getWorkspaceMembers() {
+  const child = spawn("cargo", ["metadata"], { encoding: "utf8" });
+  let data = "";
+  child.stdout.on("data", chunk => data += chunk);
+  await new Promise((resolve, reject) => {
+    child.on("close", resolve);
+    child.on("error", reject);
+  });
+  const metadata = JSON.parse(data);
+  return metadata.workspace_members.map(m => {
+    const matches = m.match(/#(.*)@\d+\.\d+\.\d+$/)
+          ||  m.match(/([^/]*)#\d+\.\d+\.\d+$/);
+    if (!matches) throw new Error("failed to parse crate name from: " + m);
+    return matches[1];
+  });
 }
 
-function myFilter(item) {
-  if (item.isa && names.includes(`cranelift/codegen/src/isa/${item.isa}`)) {
-    return true;
+/// For each given target configuration, shard the workspace's crates into
+/// buckets across that config.
+async function shard(configs) {
+  const members = await getWorkspaceMembers();
+  const buckets = Array.from({ length: GENERIC_BUCKETS }, _ => new Set());
+
+  let i = 0;
+  for (const crate of members) {
+    if (SINGLE_CRATE_BUCKETS.indexOf(crate) != -1) continue;
+    buckets[i].add(crate);
+    i = (i + 1) % GENERIC_BUCKETS;
   }
-  if (item.filter && commits.includes(`prtest:${item.filter}`)) {
-    return true;
+
+  for (crate of SINGLE_CRATE_BUCKETS) {
+    buckets.push(new Set([crate]));
   }
 
-  // If any runtest was modified, re-run the whole test suite as those can
-  // target any backend.
-  if (names.includes(`cranelift/filetests/filetests/runtests`)) {
-    return true;
+  const sharded = [];
+  i = 0;
+  for (const bucket of buckets) {
+    for (const config of configs) {
+      sharded.push(Object.assign(
+        {},
+        config,
+        {
+          name: `${config.name} (${Array.from(bucket).join(', ')})`,
+          // We run tests via `cargo test --workspace`, so exclude crates that
+          // aren't in this bucket, rather than naming only the crates that are
+          // in this bucket.
+          bucket: members
+            .map(c => bucket.has(c) ? `--package ${c}` : `--exclude ${c}`)
+            .join(" "),
+          // Filter the extra features to just those that are within this bucket,
+          // otherwise cargo complains.
+          extra_features: Array.from(
+            Object.keys(config.extra_features)
+              .filter(c => bucket.has(c))
+              .map(c => Array.from(
+                config.extra_features[c].map(f => `--features ${c}/${f}`)
+              ).join(" ")))
+            .join(" "),
```

Comment:
AFAIK the `extra_features` bits here are exclusively for wasi-nn and ONNX stuff so I feel like we're carrying quite a lot of stuff in CI configuration just for that. At that point it might be best to just split out that build/test entirely into its own job? That'd help remove all the special casing around `extra_features` here